In [2]:
import pandas as pd 
file = pd.read_csv("bankloans.csv")
file.head()

age  ed  employ  address  income  debtinc   creddebt   othdebt  default
0   41   3      17       12     176      9.3  11.359392  5.008608      1.0
1   27   1      10        6      31     17.3   1.362202  4.000798      0.0
2   40   1      15       14      55      5.5   0.856075  2.168925      0.0
3   41   1      15       14     120      2.9   2.658720  0.821280      0.0
4   24   2       2        0      28     17.3   1.787436  3.056564      1.0

In [ ]:
file.info()

<class 'pandas.DataFrame'>
RangeIndex: 1150 entries, 0 to 1149
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1150 non-null   int64  
 1   ed        1150 non-null   int64  
 2   employ    1150 non-null   int64  
 3   address   1150 non-null   int64  
 4   income    1150 non-null   int64  
 5   debtinc   1150 non-null   float64
 6   creddebt  1150 non-null   float64
 7   othdebt   1150 non-null   float64
 8   default   700 non-null    float64
dtypes: float64(4), int64(5)
memory usage: 81.0 KB


In [ ]:
file.describe()

age           ed       employ      address       income  \
count  1150.000000  1150.000000  1150.000000  1150.000000  1150.000000   
mean     35.235652     1.695652     8.781739     8.485217    47.982609   
std       8.089961     0.927051     6.914762     6.977725    40.508814   
min      20.000000     1.000000     0.000000     0.000000    13.000000   
25%      29.000000     1.000000     3.000000     3.000000    24.000000   
50%      35.000000     1.000000     7.000000     7.000000    36.000000   
75%      41.000000     2.000000    13.000000    12.000000    56.750000   
max      56.000000     5.000000    33.000000    34.000000   446.000000   

           debtinc     creddebt      othdebt     default  
count  1150.000000  1150.000000  1150.000000  700.000000  
mean     10.063391     1.605111     3.103844    0.261429  
std       6.584288     2.135967     3.529273    0.439727  
min       0.100000     0.011696     0.045584    0.000000  
25%       5.200000     0.415584     1.047996    0.000000  
50%       8.750000     0.899130     2.038053    0.000000  
75%      13.600000     1.898820     3.864960    1.000000  
max      41.300000    20.561310    35.197500    1.000000

In [ ]:
file['default'].value_counts(normalize=True)*100

default
0.0    73.857143
1.0    26.142857
Name: proportion, dtype: float64

In [4]:
file.groupby('default').mean()

age        ed    employ   address     income    debtinc  \
default                                                                  
0.0      35.514507  1.659574  9.508704  8.945841  47.154739   8.679304   
1.0      33.010929  1.901639  5.224044  6.393443  41.213115  14.727869   

         creddebt   othdebt  
default                      
0.0      1.245493  2.773409  
1.0      2.423865  3.862807

Non-Defaulter (0):
average age : 35.5 yrs 
average income : 47.154
Longer employment duration 
Lower debt to income 
Lower credit debt and other debt 

Defaulter (1):
Lower age 
Shorter employment duration 
Average income : 41.213 
High debt to income 
Lower credit debt and other debt

In [11]:
file.corr()['default'].sort_values()

employ     -0.282978
address    -0.164451
age        -0.137657
income     -0.070970
ed          0.114676
othdebt     0.145713
creddebt    0.244740
debtinc     0.389575
default     1.000000
Name: default, dtype: float64

correlation analysis shows that 
debt - to - income is (0.389) is the strongest predictor of default.

employment duration (-0.283) shoes that borrowers with longer employment history are less likely to default.

credit debt (0.244) also significantly increase risk.

this suggest financial stress indicators are stronger default predictors than demographic variables like age or education.

In [12]:
file.groupby('default')['debtinc'].describe()

count       mean       std  min   25%   50%    75%   max
default                                                          
0.0      517.0   8.679304  5.615197  0.4  4.50   7.3  11.80  32.5
1.0      183.0  14.727869  7.902798  0.9  8.55  13.8  19.35  41.3

Debt to income ratio analysis
 the avg. debt to income ratio for defaulters (14.74) is significantly higher than non defaulters (8.68)

 non defaulters median : 7.3
 defaulters median : 13.8

 this suggest that debt to income ratio is a strong predictor of default risk.

 a potential risk threshold appears around 12-13,
 above which default probability increases substantially


In [4]:
file['risk_level'] = pd.cut(file['debtinc'], bins=[0,8,12,50], labels=['low','medium','high'])
file.groupby(['risk_level', 'default']).mean()

age        ed    employ    address     income  \
risk_level default                                                        
low        0.0      35.076125  1.688581  9.477509   8.411765  49.013841   
           1.0      31.400000  1.975000  3.800000   5.450000  41.650000   
medium     0.0      35.836538  1.586538  9.076923  10.182692  42.336538   
           1.0      31.468750  1.875000  4.625000   4.468750  36.968750   
high       0.0      36.266129  1.653226  9.943548   9.153226  46.862903   
           1.0      34.036036  1.882883  5.909910   7.288288  42.279279   

                      debtinc  creddebt   othdebt  
risk_level default                                 
low        0.0       4.727682  0.753630  1.601505  
           1.0       5.145000  0.908108  1.410192  
medium     0.0       9.980769  1.318284  2.938601  
           1.0       9.806250  1.447157  2.154437  
high       0.0      16.797581  2.330799  5.366153  
           1.0      19.600000  3.251657  5.239136

risk segmentation based on debt to income ratio 

default rate increases significantly across risk buckets :

- low risk (<8>):12% default rate
- medium risk (8-12): 23% default rate 
- high risk (>12): 47% default rate 

this shows a strong monotonic relationship between debt to income ratio and default probability .
debt - to - income ratio can be used as a primary screening variable in credit underwriting


In [6]:
risk_table = file.groupby("risk_level")['default'].mean()
risk_table *100

risk_level
low       12.158055
medium    23.529412
high      47.234043
Name: default, dtype: float64

 Probability of Default by Risk Segment

Customers were segmented based on debt-to-income ratio.

Probability of Default (PD):

- Low Risk: 12.16%
- Medium Risk: 23.53%
- High Risk: 47.23%

There is a strong monotonic increase in default probability 
as debt-to-income ratio increases.

This confirms that debt-to-income ratio is a key driver 
of credit risk and should be considered in underwriting decisions.

Customers with DTI > 12 show sharply increasing default risk.
High-risk segment (~47% PD) should require stricter underwriting.
Medium-risk customers could require additional income verification.
Low-risk customers are safer for standard loan approval.